### ISSUE TITLE
Custom validation set size

### ISSUE THREAD

**Issue by ecly** [modernmt/modernmt#524]
For very large datasets, the current ~3000 extracted validation pairs can be quite unrepresentative, and it would be great if we could default to a percentage, and possibly take the amount of validation pairs as an argument, either integer or percentage.

**Comment by nicolabertoldi** [comment]
@ecly 

we will take into consideration your suggestion and discuss about it whether to implement it in future releases.

By the way, when training data is really huge (+100M sentences) even a devset 10 times bigger (30K sentences) could be not representative if randomly selected.
Otherwise, if you "manually" select a devset there is always the risk to focus the engine towards a specific domain

Finally, if you want to check whether increasing the size of the development set improves the overall performance of the engine you can modify the hardcoded size here (https://github.com/modernmt/modernmt/blob/8d444da2e8ac77318901a9d357c3d500ad5c6059/cli/datagen.py#L68)

A deep investigation on this issue would be interesting. If you did something in this direction, we would appreciate you feedback.

**Comment by ecly** [comment]
@nicolabertoldi 

I see your point and agree. Lots of different factors to consider.

I have actually been using 30,000 as devset size myself (by modifying the hardcoded size) for model training with anything between ~10M and ~100M sentences, and what I found is that the validation results (and thereby also the timing of early stopping) became vastly more consistent both across language combinations, and across different training runs on the same language pair (with growing datasets).

**Comment by nicolabertoldi** [comment]
@ecly 

your results are really interesting.
Thanks for sharing with us.

**Comment by EtienneAb3d** [comment]
On my side, I prefer to consider that the validation set is never pertinent. It's even more a fact that I explicitly deactivate the pre-filtering process to keep doubles. I'm doing this for few reasons, an important one is to be able to artificially enlarge in-domain data by duplicating them several times. Thus, it's highly probable that the validation set contains doubles from the training set, producing a validation estimation decreasing in the same kind of way as the training estimation.
My logic is to ensure that the training data are seen "several times". For 100M data set, doing 10 full epochs means about 2 weeks of training on a single 2080.
